hoomd/hpmc/PairPotentialUnion.cc

```diff
@@ -0,0 +1,292 @@
+                {
+                throw std::length_error("orientations must be a list of 4-tuples.");
+                }
+
+            m_orientation[body_type_id][i]
+                = quat<LongReal>(q_python[0].cast<LongReal>(),
+                                 vec3<LongReal>(q_python[1].cast<LongReal>(),
+                                                q_python[2].cast<LongReal>(),
+                                                q_python[3].cast<LongReal>()));
+            }
+
+        m_type[body_type_id][i]
+            = m_sysdef->getParticleData()->getTypeByName(types[i].cast<std::string>());
+
+        if (charges.is_none())
+            {
+            m_charge[body_type_id][i] = 0;
+            }
+        else
+            {
+            m_charge[body_type_id][i] = pybind11::list(charges)[i].cast<LongReal>();
+            }
+        }
+
+    updateExtent(body_type_id);
+    buildOBBTree(body_type_id);
+    notifyRCutChanged();
+    }
+
+pybind11::object PairPotentialUnion::getBody(std::string body_type)
+    {
+    unsigned int body_type_id = m_sysdef->getParticleData()->getTypeByName(body_type);
+    size_t N = m_position[body_type_id].size();
+    if (N == 0)
+        {
+        return pybind11::none();
+        }
+
+    pybind11::list positions;
+    pybind11::list orientations;
+    pybind11::list types;
+    pybind11::list charges;
+
+    for (unsigned int i = 0; i < N; i++)
+        {
+        positions.append(pybind11::make_tuple(m_position[body_type_id][i].x,
+                                              m_position[body_type_id][i].y,
+                                              m_position[body_type_id][i].z));
+        orientations.append(pybind11::make_tuple(m_orientation[body_type_id][i].s,
+                                                 m_orientation[body_type_id][i].v.x,
+                                                 m_orientation[body_type_id][i].v.y,
+                                                 m_orientation[body_type_id][i].v.z));
+        types.append(m_sysdef->getParticleData()->getNameByType(m_type[body_type_id][i]));
+        charges.append(m_charge[body_type_id][i]);
+        }
+
+    pybind11::dict v;
+    v["types"] = types;
+    v["positions"] = positions;
+    v["orientations"] = orientations;
+    v["charges"] = charges;
+    return std::move(v);
+    }
+
+void PairPotentialUnion::updateExtent(unsigned int type_id)
+    {
+    auto N = static_cast<unsigned int>(m_position[type_id].size());
+
+    Scalar extent_i = 0.0; // 2x the dist of farthest-away constituent particle of type i
+
+    for (unsigned int i = 0; i < N; i++)
+        {
+        auto pos = m_position[type_id][i];
+
+        // extent_i is twice the distance to the farthest particle, so is ~ the circumsphere
+        // diameter
+        Scalar r = sqrt(dot(pos, pos));
+        extent_i = std::max(extent_i, Scalar(2 * r));
+        }
+
+    // set the diameter
+    m_extent_type[type_id] = extent_i;
```

Comment:
```suggestion
    m_extent_type[type_id] = std::transform_reduce(
        m_position[type_id].begin(),
        m_position[type_id].end(),
        0.0,
        std::max,
        [](auto& pos) { return 2 * sqrt(dot(pos, pos)); }
    )
```
Might need some tweaking to work.

Reply:
This does not work. The compiler cannot determine which template of std::max to use. I instead shortened the code with a range-based for loop.